### ISSUE TITLE
The useless attribute lint doesn't put the ! in the right place.

### ISSUE THREAD

**Issue by BookOwl** [Manishearth/rust-clippy#1522]
When the useless_attribute lint is activated, it says that you most likely forgot an exclamation mark (!). While this is true, it also suggests to stick the ! in the middle of the attribute:
```
warning: useless lint attribute, #[warn(useless_attribute)] on by default
 --> src/lib.rs:1:38
  |
1 | #[cfg_attr(feature = "cargo-clippy", allow(new_without_default))]
  |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
help: if you just forgot a `!`, use
  | #[cfg_attr(feature = "cargo-clippy", a!llow(new_without_default))]
  = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#useless_attribute
```
The correct response should be to suggest adding it after the #

**Comment by Cyres** [comment]
I just had this issue with `enum_glob_use` too. I guess, the message is bugged in general.

**Comment by phansch** [comment]
It looks like there is no attribute information for `cfg_attr` currently (`item.attrs` does not include it at all). I will take a stab at this and see if I can find a workaround.

**Comment by phansch** [comment]
Unfortunately this seems really difficult to find a proper fix for. Since `cfg_attr` and `cfg` attributes are not accessible in the linting process, the only way to work around that, would be to use the CodeMap to search for the beginning of the attribute.

I managed to get a fix working, but only if the whole `cfg` attribute is on a single line. Cases where the `cfg` attribute is split over multiple lines would still be broken. I'm not sure if it makes sense to parse the whole file ourselves and look for the beginning of each attribute that spans multiple lines.

Maybe it's easier to remove the suggestion from this lint for now? @oli-obk

**Comment by oli-obk** [comment]
@phansch can you detect the cfg case? if so, just don't emit the suggestion in the cfg case

**Comment by phansch** [comment]
Sorry, my previous comment was partly wrong (now updated):

As far as I can tell, only `cfg_attr` is missing from the item attributes, so I think we don't have to worry about `cfg`. Using CodeMap, I can detect `cfg_attr` somewhat easily if it's on the same line, but it's more work to detect if the attribute starts on a previous line, like

```rust
#[cfg_attr(feature = "cargo-clippy",
  allow(dead_code),
  allow(new_without_default))]
```
That would require iterating over each line towards the beginning of the file, making a Span out of each Line, checking if the span contains the beginning of the attribute (a `#[`), if yes, replacing it with `#![`.

`cfg` is included in the item attributes, but I'm not sure if it matters for this lint, because as it is right now, `cfg` attributes are not linted in this lint.

**Comment by oli-obk** [comment]
So... we could just not create a suggestion at all in the multi line case?

**Comment by phansch** [comment]
That would be the easier solution, yes. If the attribute starts on the same line, we show the correct suggestion, otherwise, we don't show a suggestion 👍
